dist/tools/buildsystem_sanity_check/check.sh

```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env bash
+#
+# Copyright (C) 2018 Gaëtan Harter <gaetan.harter@fu-berlin.de>
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+#
+
+#
+# Central test script to have sanity checks for the build system
+# It is run unconditionally on all files.
+#
+#
+
+: "${RIOTBASE:="$(cd "$(dirname "$0")/../../../" || exit; pwd)"}"
+
+SCRIPT_PATH=dist/tools/buildsystem_sanity_check/check.sh
+
+# Modules should not check the content of FEATURES_PROVIDED/_REQUIRED/OPTIONAL
+# Handling specific behaviors/dependencies should by checking the content of:
+# * `USEMODULE`
+# * maybe `FEATURES_USED` if it is not a module (== not a periph_)
+check_not_parsing_features() {
+    local patterns=()
+    local pathspec=()
+
+    patterns+=(-e 'if.*filter.*FEATURES_PROVIDED')
```

Comment:
using `if[^[].*filter.*FEATURES_PROVIDED` should be equally effective and not match itself, thus removing the need for line 18 and 36.

```suggestion
    patterns+=(-e 'if[^[].*filter.*FEATURES_PROVIDED')
```

Off topic, this thing of regexes that match themselves seems interesting.

Reply:
The problem is not the regex in particular. The file should not be checked, like Makefile.include

If later other regex, or documentation are introduced then they could also match.

I could update the documentation about the "self match".